[Merged by Bors] - asset cards with more metadata

IceSentry opened the pull request:
## Objective

Update the asset cards to support more metadata. The main goal is to be able to show the supported bevy version. Everything else is secondary right now.

## Solution

Support more metadata by using the styling from https://github.com/bevyengine/bevy-website/issues/231.

I'm missing a few things, like different colors for bevy versions, but that's because I don't know how to do that with zola templates. I also ignored the author name and crate version for now because it doesn't apply to every crate and it can be iterated on in the future.

![image](https://user-images.githubusercontent.com/8348954/179620589-77b196df-c0ed-412f-ac78-8876e3430f65.png)

Supported metadata:
- `bevy_versions`: array of supported version
- `licenses`: array of licenses

The metadata currently doesn't exist, but the code is written in such a way to not break if it doesn't exist. I created #397 to generate the required metadata.

A lot of the values were just found by me playing around with values and finding something that looked good. I'm absolutely opened to changing them. To be clear, a lot of what I have right now are just values that ended up working, but there's probably a better way to do most of those things.

## Remaining questions

- ~~What to do with assets without images.~~
  - Solved: I added a gradient to make it a bit less empty.
- How to handle different color based on the value of the version tag by using zola templates

## Notes

Until we add the required metadata, the cards will have a lot of empty space at the bottom, I already have a prototype ready to add the missing metadata, so that should be fixed pretty fast.

IceSentry commented:
@TimJentzsch since I can't figure out how to request a review, hopefully this ping will do the job.

mockersf commented on sass/components/_card.scss:
```diff
@@ -85,4 +85,4 @@
     margin-right: auto;
     align-items: center;
-}
+}
```
```suggestion
}

```

mockersf commented:
How it looks for me (Safari and Firefox):
<img width="953" alt="Screenshot 2022-07-18 at 12 06 49" src="https://user-images.githubusercontent.com/8672791/179489820-91e844fa-1114-4b64-bc7b-f080dcaa9685.png">

* The icons seems to not be aligned with the centre of the tags
* The license icon is too close to the license tag
* If there are no version, the license is on the left

There is also an issue when there are too many bevy versions, this would hide the licenses. That could be solved later as it's unlikely to be an issue today

For the remaining questions, I would like this to move without solving them:
* Assets without an image: let's keep them empty for now, and take the time to find a good placeholder
* Colors for versions: colors are... too complicated. This info is needed now, even without colors

TimJentzsch commented:
> - The icons seems to not be aligned with the centre of the tags
> - The license icon is too close to the license tag
> - If there are no version, the license is on the left

We should be able to fix all of that with :sparkles: flexbox :sparkles: 

The main ingredients:

- `display: flex`: This is where the fun begins!
- `justify-content: space-between`: The first item goes to the start, the second item goes to the end.
- Wrap the version and license each in a `div`. Only the content of the `div` is conditional, that way you keep the license to the right.
- `align-items: center`: Aligns the icons at the center of the tags.
- `gap: 5px`: Space between icon and tag name.
- `flex-flow: row-wrap`: If there are too many bevy versions, the license wraps in the next row.

That should be everything we need. In a couple of hours I might also have time to do a mockup with vanilla HTML and CSS if needed.

afonsolage commented:
Since the new layout always expects an image (like a thumbnail), maybe add a [default bevy image](https://bevyengine.org/processed_images/4e424e8bfdfe1d4800.png) when there is no image on metadata?

![image](https://user-images.githubusercontent.com/1176452/179522861-254bc315-3064-4e5a-a905-d4ffb6c492d6.png)

The current layout adapts when there is no image.

mockersf commented:
> Since the new layout always expects an image (like a thumbnail), maybe add a [default bevy image](https://bevyengine.org/processed_images/4e424e8bfdfe1d4800.png) when there is no image on metadata?

Yes, we'll need a placeholder, but not the Bevy logo

TimJentzsch commented:
> The current layout adapts when there is no image.

In my opinion that makes more sense than having a big gray box that is not used

IceSentry commented:
For the image thing, I was mostly questioning what to do right now, but yes in the future we should either make images mandatory or at the very least some sort of generated placeholder. I believe once cards are more clearly used to show case images we'll see more people submitting images. Could be worth explicitly calling out in the blog post.

IceSentry commented:
For the bevy and license logo, that was me playing around with it and breaking it after making my original PR it was super late though so I stopped and went to bed. I'll try to fix it later today.

doup commented:
It's looking good!

For the image adding some light background image (maybe some grey cogs?) + the name of the asset on top should look good enough. I think that removing it, with this vertical layout, is going to look ugly as the title/text will be misaligned between cards.

I have a couple of minor nits on margins/padding/alignment. On top of what `@mockersf` already mentioned in https://github.com/bevyengine/bevy-website/pull/394#issuecomment-1187016606, I would say that the margins/padding would look better if they were more consistent:
- Reduce bottom gap of the card (on the current online version on both assets/examples cards the same gap is used for both axis). You probably have some `margin-bottom`; just removing it should work as the parent `grid` container is responsible for the gap.
- Increase left padding on asset title to match description
- Increase bottom padding for de card

![image](https://user-images.githubusercontent.com/188612/179573776-7df84958-fc31-4607-8f8b-40b25759a263.png)

Also, on your screenshot it looks like the text is not vertically centered on the "pills":

![image](https://user-images.githubusercontent.com/188612/179574500-2d07a49c-338d-43ed-bb3b-bb12cd11e479.png)

IceSentry commented:
Alright, I rewrote most of the styling, but this time I used css grid and it made everything way easier. I tried to apply most of the feedback that I saw here.

I also added a gradient for the assets without images just so it's less empty

![image](https://user-images.githubusercontent.com/8348954/179580805-ab5b3462-179d-4a62-a5bc-7a51166f3ca6.png)

@doup sorry I didn't see your feedback before committing, but I think I fixed all of that

TimJentzsch commented on sass/pages/_assets.scss:
```diff
@@ -90,22 +90,34 @@
-            color: $link-color;
-        }
     }
 }
```
Nit: Can we add a trailing linebreak here? GitHub complains about that

IceSentry replied:
Is it possible you didn't refresh? My latest commit should have already fixed that one

TimJentzsch commented on templates/macros/assets.html:
```diff
@@ -1,56 +1,94 @@
+        <div class="asset-card__banner">
+            <div class="asset-card__title">{{post.title}}</div>
+            {% if post.extra.image %}
             <img src="{{ image_macros::resize_image(path=post.extra.image, width=340, height=340) }}" />
```
This seems to be missing an `alt` attribute`

TimJentzsch commented on templates/macros/assets.html:
```diff
@@ -1,56 +1,94 @@
+            {% endif %}
+            {% if post.extra.licenses %}
+            <div class="asset-card__tag-list asset-card__licenses">
+                {{ assets_macros::license_logo_svg() }}
+                {% for license in post.extra.licenses %}
+                <span class="asset-card__tag">{{license}}</span>
```
Similar here, except that `alt` attributes don't work on `svg`s.
From what I've read, the best way to do this is to add a `title` element inside the `svg`: [Blog post](https://css-tricks.com/accessible-svgs/#aa-2-inline-svg), [W3 Reference](https://www.w3.org/TR/SVG11/struct.html#DescriptionAndTitleElements)

IceSentry replied:
I want to convert the svg to an image asset like the bevy logo anyway so that will work

TimJentzsch commented on templates/macros/assets.html:
```diff
@@ -1,56 +1,94 @@
+    {{ self::assets_menu_row(prefix=prefix, section_path=s)}}
+    {% endfor %}
+</ul>
+{% endmacro %}
```
Nit: Trailing line break is missing here

TimJentzsch commented on templates/macros/assets.html:
```diff
@@ -1,56 +1,94 @@
-                {% endfor %}
-            </ul>
+        <label class="tree-menu__toggle" for="{{id}}">
+            <img class="tree-menu__chevron" src="/assets/icon-chevron-down.svg">
```
This is missing an `alt` attribute and should maybe have a `title` attribute to indicate that you can toggle something here (if I understand this correctly)

TimJentzsch reviewed (changes requested):
The screenshots you sent look good to me!

I did a review run on accessibility issues and found a few, would be nice to fix those before merging :)

IceSentry commented on templates/macros/assets.html:
```diff
@@ -1,56 +1,94 @@
+        <div class="asset-card__tags">
+            {% if post.extra.bevy_versions %}
+            <div class="asset-card__tag-list asset-card__bevy-versions">
+                <img src="/assets/bevy_icon_dark.svg" alt="Small bevy icon">
+                {% for version in post.extra.bevy_versions %}
+                <span class="asset-card__tag">{{version}}</span>
```
I literally had that exact "Supported bevy versions" alt text earlier, I have no idea why I changed it lol and yeah, using a title is a good call.

TimJentzsch commented on templates/macros/assets.html:
```diff
@@ -37,7 +24,7 @@
             {% if post.extra.bevy_versions %}
             <div class="asset-card__tag-list asset-card__bevy-versions">
-                <img src="/assets/bevy_icon_dark.svg" alt="Small bevy icon">
+                <img src="/assets/bevy_icon_dark.svg" title="Supported bevy versions">
```
Nit: we need both the `alt` and `title`, as the `alt` is also rendered when the `img` can't be loaded.

(Sorry, I should have made that clearer in my comment)

TimJentzsch commented on templates/macros/assets.html:
```diff
@@ -71,7 +58,8 @@
         {% if section.subsections %}
         <label class="tree-menu__toggle" for="{{id}}">
-            <img class="tree-menu__chevron" src="/assets/icon-chevron-down.svg">
+            <img class="tree-menu__chevron" src="/assets/icon-chevron-down.svg"
+                title="Toggle {{section.title}} section">
```
Same here, we also need the `alt` attribute.

TimJentzsch commented on templates/macros/assets.html:
```diff
@@ -21,14 +8,14 @@
-
 {% macro card(post) %}
 <div id="{{ post.title | slugify }}">
     <a class="asset-card" href="{{ post.extra.link }}">
```
The [MDN docs for `a`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a) state:

> Content within each `<a>` **should** indicate the link's destination.

Such that users with visual impairments/using screen-readers can tell where they are going.
I assume that this is not the case here, because we are wrapping a `div` element.

However, I could not find any guidelines to make this accessible. If anyone knows anything, please let me know!

Maybe it's fine though because of the card title that's deeper in the hierarchy?

TimJentzsch replied:
I think they mean that the content of the `<a>` should indicate where the link leads to: [Docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#strong_link_text)

TimJentzsch reviewed (approved):
Awesome, looks good to me!

doup commented on sass/pages/_assets.scss:
```diff
@@ -18,29 +18,109 @@
+            }
+        }
+
+        &__title {
+            position: absolute;
+            width: 100%;
+            bottom: 0;
+
+            padding: 0.3rem;
+            font-weight: bold;
+            font-size: 1.2rem;
+            background-color: rgba(0, 0, 0, 0.5);
+        }
+
+        &__description {
+            grid-area: desc;
+            display: flex;
+
+            padding: 0.3rem;
+            font-size: 1rem;
+        }
```
I've tried with a little bit of extra padding on the title/description (0.6rem ~ 8px), what do you think?

```suggestion
        &__title {
            position: absolute;
            width: 100%;
            bottom: 0;

            padding: 0.6rem;
            font-weight: bold;
            font-size: 1.2rem;
            background-color: rgba(0, 0, 0, 0.5);
        }

        &__description {
            grid-area: desc;
            display: flex;

            padding: 0.6rem;
            font-size: 1rem;
        }
```

![image](https://user-images.githubusercontent.com/188612/179683171-c5f2c4ac-12da-4ab7-856c-a97317067b8a.png)

doup replied:
Sure, go for it. `0.6` was a suggestion.

doup commented on sass/pages/_assets.scss:
```diff
@@ -18,29 +18,109 @@
-        margin: 0;
-        height: 8rem;
-    }
+    .asset-card {
```
I would move `asset-card` into it's own file `sass/components/_asset-card.scss`.
Pages folder (`pages/*`) is kind-of legacy; where both BEM and non-BEM gets mixed.

IceSentry replied:
Sure, I'm still new to BEM so wasn't sure how file structure should work so I just updated the existing file. I'll do that tomorrow.

doup commented on sass/pages/_assets.scss:
```diff
@@ -18,29 +18,109 @@
+            width: 100%;
+            bottom: 0;
+
+            padding: 0.6rem;
```
This shrinks the vertical padding for the title, this way it doesn't look so "airy"… and I stop. 😇 
Anyway, if you think `0.6rem` is too much please reduce as you wish… it's just that I find `0.3rem` too tight.

```suggestion
            padding: 0.3rem 0.6rem;
```

![image](https://user-images.githubusercontent.com/188612/179687507-c93c7833-0f51-4e93-8b5a-30d5f699e02d.png)

IceSentry replied:
😍 I love it, I that was the missing part.

doup commented on sass/pages/_assets.scss:
```diff
@@ -18,29 +18,109 @@
+            background-color: rgba(0, 0, 0, 0.5);
+        }
+
+        &__description {
+            grid-area: desc;
+            display: flex;
+
+            padding: 0.6rem;
+            font-size: 1rem;
+        }
```
I've noticed that this doesn't handle extra long text correctly.

**Option A**: use 4 lines ellipsis ([the mixin uses this snippet](https://css-tricks.com/almanac/properties/l/line-clamp/)).

```suggestion
        &__description {
            @include text-ellipsis(4);

            grid-area: desc;

            padding: 0.6rem;
            font-size: 1rem;
        }
```

**Option B**, see comment:

doup replied:
Sorry, I didn't realize, I saw it a after you removed it again on the comment. 🤦

doup commented on sass/pages/_assets.scss:
```diff
@@ -18,29 +18,109 @@
+            "banner"
+            "desc"
+            "tags";
+        grid-template-rows: 12rem 5.5rem 2rem;
```
**Option B**, for comment: https://github.com/bevyengine/bevy-website/pull/394/files#r924157060

```suggestion
        grid-template-rows: 12rem minmax(5.5rem, 1fr) 2rem;
        height: 100%;
```

- `minmax` to allow the description grow, but keep a minimum height
- `height` for the other cards, so that they grow to match the one with the long description

![image](https://user-images.githubusercontent.com/188612/179692203-4e1c381b-aeab-404e-8f27-9ccdbeea0c8c.png)

ickk replied:
I'm biased against growing the cards vertically; when we were designing the initial mockups we had a lot of discussion about the amount of text that would be appropriate for cards, and we landed on the ratios and everything that we did based on a relatively little amount of text. 

Descriptions should be short and to the point. Longer exposition about a particular asset belongs on it's asset page (whenever we build that feature) or on the crates.io/repo/webpage linked to imo

doup replied:
This was just an option I was giving him, related comment: https://github.com/bevyengine/bevy-website/pull/394/files#r924157060
He used **"Option A"** in the end, which is just "ellipsis at the fourth line": https://github.com/bevyengine/bevy-website/pull/394/commits/1a93da13e2cdbe0d8d8b2ac4949dba917e76f775

alice-i-cecile commented:
@IceSentry this needs merge conflict resolution, but IMO we should ship this and iterate.

IceSentry commented:
I'd rather shipping the metadata loading first. That way the cards will have actual data to show. Right they will just look weirdly empty

alice-i-cecile commented:
Sounds good to me. I'll mark this as Blocked so we don't forget.

IceSentry commented:
Well, technically it isn't blocked, it could easily be merged without any issue, it would just look a bit empty at the bottom since there would be no license or bevy versions

alice-i-cecile commented:
Technically, but I'd be against it :p

IceSentry commented:
The other PR is ready it just needs more reviews. it's just that most people don't look at `bevy-website` 😦

cart commented on sass/components/_asset-card.scss:
```diff
@@ -0,0 +1,107 @@
+        display: block;
+        position: relative;
+
+        background: linear-gradient(45deg,
```
Totally a matter of opinion, but I'd like to propose a more subtle / aligned gradient: `linear-gradient(0deg, #37393e 0%, #1b1b1b 100%)`

![image](https://user-images.githubusercontent.com/2694663/185258258-d8741f68-0156-4084-afd4-3c6526e3a9af.png)

The current one draws the eye a bit too much and feels a bit aggressive / like a bright light is shining on something very metallic / reflective. And notably with that analogy the light should behave differently for each tile based on its position relative to your eye, but it doesn't. Which also contributes to my (personal) impression that it feels "off":

![image](https://user-images.githubusercontent.com/2694663/185258410-747669b3-d8ee-4fb5-9be2-8674fd31a142.png)

IceSentry replied:
Good idea, I like it, a lot more subtle while still being nicer than a solid color.

cart commented on sass/components/_asset-card.scss:
```diff
@@ -0,0 +1,107 @@
+        "banner"
+        "desc"
+        "tags";
+    grid-template-rows: 12rem 5.5rem 2rem;
```
I think we should probably rein in the height of the text section a bit (cut down on the character limit if necessary, maybe reduce the font size). The cards have _just_ a bit too much height right now. Most descriptions are short and this results in _a lot_ of wasted space. I'm down to do that in a separate PR though.

IceSentry replied:
I'll see what I can do. I picked the height based on the longer descriptions but it was mostly just playing with some values. I'm down with reducing the size a bit

IceSentry replied:
I reduced the padding a bit and limited it to 3 lines now. This is how it looks.
![image](https://user-images.githubusercontent.com/8348954/185264158-c7c35120-c9a8-4a69-b945-a030350ea1f6.png)

I tried going to 2 lines max but it looked weird.

cart replied:
I dig it! One more small tweak: I do think adding back a bit more padding is "worth it".

IceSentry replied:
I increased the padding a bit and made the area for the tags slightly larger so it could have a more consistent padding.

![image](https://user-images.githubusercontent.com/8348954/185279660-f7f270b5-21ff-40c6-a336-38e190445deb.png)

cart replied:
Lovely!

cart reviewed (commented):
I'm loving these changes!

rparrett commented:
I think that we may have lost lazy-loading of the asset images in the rebase.

rparrett commented on templates/macros/assets.html:
```diff
@@ -1,56 +1,93 @@
+        <div class="asset-card__banner">
+            <div class="asset-card__title">{{post.title}}</div>
+            {% if post.extra.image %}
+            <img src="{{ image_macros::resize_image(path=post.extra.image, width=340, height=340) }}"
```
```suggestion
            <img src="{{ image_macros::resize_image(path=post.extra.image, width=370, height=370) }}"
```

cart commented on templates/macros/assets.html:
```diff
@@ -1,11 +1,16 @@
+<svg style="display: none" version="2.0">
+    <defs>
+        <symbol id="asset-card__scale-balanced-svg" viewBox="0 0 640 512">
+            <!--! Font Awesome Pro 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license (Commercial License) Copyright 2022 Fonticons, Inc. -->
```
Hmm this is explicitly calling itself out as "font awesome pro". We shouldn't be using any icons from there (and I dont think this one is part of pro).

cart replied:
https://fontawesome.com/license/free

IceSentry replied:
That's the exact svg with the same comment I get when I copy it from there https://fontawesome.com/icons/scale-balanced?s=solid

I'm not sure why it says it's under the pro license

cart replied:
If you download the "free" package, the svgs have the correct license embedded in them. I just pushed that svg content.

IceSentry replied:
Interesting, it's a bit weird that they have the wrong license on the site for the exact same svg. That's good to know though so I downloaded it for the future.

cart commented:
bors r+
